src/thread.rs

```diff
@@ -0,0 +1,483 @@
+            threads: threads,
+            blockset_counter: 0,
+            thread_local_alloc_ids: Default::default(),
+            yield_active_thread: false,
+        }
+    }
+}
+
+impl<'mir, 'tcx: 'mir> ThreadManager<'mir, 'tcx> {
+    /// Check if we have an allocation for the given thread local static for the
+    /// active thread.
+    fn get_thread_local_alloc_id(&self, def_id: DefId) -> Option<AllocId> {
+        self.thread_local_alloc_ids.borrow().get(&(def_id, self.active_thread)).cloned()
+    }
+
+    /// Set the allocation id as the allocation id of the given thread local
+    /// static for the active thread.
+    fn set_thread_local_alloc_id(&self, def_id: DefId, new_alloc_id: AllocId) {
+        assert!(
+            self.thread_local_alloc_ids
+                .borrow_mut()
+                .insert((def_id, self.active_thread), new_alloc_id)
+                .is_none(),
+            "a thread local initialized twice for the same thread"
+        );
+    }
+
+    /// Borrow the stack of the active thread.
+    fn active_thread_stack(&self) -> &[Frame<'mir, 'tcx, Tag, FrameData<'tcx>>] {
+        &self.threads[self.active_thread].stack
+    }
+
+    /// Mutably borrow the stack of the active thread.
+    fn active_thread_stack_mut(&mut self) -> &mut Vec<Frame<'mir, 'tcx, Tag, FrameData<'tcx>>> {
+        &mut self.threads[self.active_thread].stack
+    }
+
+    /// Create a new thread and returns its id.
+    fn create_thread(&mut self) -> ThreadId {
+        let new_thread_id = ThreadId::new(self.threads.len());
+        self.threads.push(Default::default());
+        new_thread_id
+    }
+
+    /// Set an active thread and return the id of the thread that was active before.
+    fn set_active_thread_id(&mut self, id: ThreadId) -> ThreadId {
+        let active_thread_id = self.active_thread;
+        self.active_thread = id;
+        assert!(self.active_thread.index() < self.threads.len());
+        active_thread_id
+    }
+
+    /// Get the id of the currently active thread.
+    fn get_active_thread_id(&self) -> ThreadId {
+        self.active_thread
+    }
+
+    /// Has the given thread terminated?
+    fn has_terminated(&self, thread_id: ThreadId) -> bool {
+        self.threads[thread_id].state == ThreadState::Terminated
+    }
+
+    /// Get the borrow of the currently active thread.
+    fn active_thread_mut(&mut self) -> &mut Thread<'mir, 'tcx> {
+        &mut self.threads[self.active_thread]
+    }
+
+    /// Mark the thread as detached, which means that no other thread will try
+    /// to join it and the thread is responsible for cleaning up.
+    fn detach_thread(&mut self, id: ThreadId) {
+        self.threads[id].detached = true;
+    }
+
+    /// Mark that the active thread tries to join the thread with `joined_thread_id`.
+    fn join_thread(&mut self, joined_thread_id: ThreadId) {
+        assert!(!self.threads[joined_thread_id].detached, "Bug: trying to join a detached thread.");
```

Comment:
Is it truly impossible for the program to reach this branch? Looks like the shims then have to protect against that case, which means code duplication between POSIX and Windows shims. So maybe it would be better if instead this function raise a `throw_ub_format!`?